### PR TITLE
Muon Fitting Refactor

### LINKED ISSUE #29892: Muon fitting refactor
The muon fitting is very complex and needs simplifying. To refactor this we will:
- Abstract the fitting layout in the fitting tab to make it more 'general'. 
- [ ] There will be a 'basic' fitting layout which can be reused for `Model Analysis` later down the line. This will **not** have `Select Workspaces` or `Simultaneous fit over`.
- [ ] There will be a 'general' fitting layout that is used for all the normal fitting in the fitting tab. This will inherit the 'basic' layout.
- [ ] There will be a 'TFAymmetry' layout for a special kind of fitting with a normalisation factor. This will inherit the 'general' layout.

- After this abstraction and refactor we can remove the TF Asymmetry checkbox completely.
- Switching between TFAsymmetry and normal fitting can be done via a combobox. A complex looking function will no longer be used in the function tree.

The fitting tab and TF Asymmetry tab will both need to work with Single and Double Pulse data.

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR is the base branch for the refactor of the muon fitting widgets. It creates the basic and general fitting widgets which can be used for normal fitting in the fitting tab of Muon Analysis and FDA interfaces. It also creates a widget which can be used for TFAsymmetry fitting.

Tests have been added for:
- [x] Basic model, view and presenter
- [x] General model, view and presenter
- [x] TFA model, view and presenter
- [x] `FittingTabPresenterTest` and `FittingTabModelTest` have been removed.

**To test:**

Fixes #29892

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
set up a simultaneous fit
Select a global
click fit
the globals are gone

**Comment by AnthonyLim23** [comment]
the `plot raw` doesn't seem to work (you need to add some rebinned data first)

it doesnt seem to update after ticking and unticking `fit to raw`

**Comment by AnthonyLim23** [comment]
if I do a fit then change the fitting function, the function name part of the results table gets reset (we should just add to it)

**Comment by AnthonyLim23** [comment]
double pulse seems to work

**Comment by AnthonyLim23** [comment]
when adding plots to the analysis (via the grouping tab or by creating a phasequad), they seem to randomly appear and disappear from the plot

**Comment by robertapplin** [comment]
Thanks @AnthonyLim23 , should be ready for another test

**Comment by AnthonyLim23** [comment]
got a couple of visual things left now.

Lets make the fit button smaller and fixed size (at the moment it covers the text for `plot guess`)
Lets give the selector a max size, so we can always read the `select workspaces` label
When doing simultaneous, ticking a global seems to cause the function browser to scroll to the left. This is annoying. Could we just make that column narrower (I see no benefit to it being so wide)?

**Comment by AnthonyLim23** [comment]
failing tests and can we change the base branch to be a fork?

**Comment by robertapplin** [comment]
@AnthonyLim23 This has now been moved to a forked repo https://github.com/robertapplin/mantid/pull/1

**Comment by AnthonyLim23** [comment]
going to reopen for testing

File: scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py
Comment (AnthonyLim23) [inline scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py:160]: should this not check we have a waarning too?

File: scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py
Comment (AnthonyLim23) [inline scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py:116]: does this need the plot_guessto be set?

File: scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py
Comment (AnthonyLim23) [inline scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py:280]: good comment

File: scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py
Comment (AnthonyLim23) [inline scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py:372]: are these mock methods going to be used in other tests?

**Comment by AnthonyLim23** [comment]
when switching between simultaneous and single fitting, we do not want to reset the start and end x. The user may set up the fit as a single (to get ball park values) then switch to simultaneous. It would then be confusing for the range to have changed

**Comment by AnthonyLim23** [comment]
Load MUSR 62260
Add a fitting function (doesn't matter which)
Do a fit
Then use the arrows to inspect the other data sets
When you get back to the one you started with, the fit will be plotted but the parameters will be wrong
If you keep clicking the arrows the parameters will appear. However, they will be on the wrong group

File: scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py
Comment (robertapplin) [reply scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py:160]: A warning isn't displayed when there is no fit function. Its the same as in the current version

File: scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py
Comment (robertapplin) [reply scripts/test/Muon/fitting_widgets/basic_fitting/basic_fitting_presenter_test.py:116]: Yes, plot guess is set in the setup of this test

**Comment by robertapplin** [comment]
> Load MUSR 62260
> Add a fitting function (doesn't matter which)
> Do a fit
> Then use the arrows to inspect the other data sets
> When you get back to the one you started with, the fit will be plotted but the parameters will be wrong
> If you keep clicking the arrows the parameters will appear. However, they will be on the wrong group

Thanks, this has been fixed in the TF Asymmetry PR #30772

**Comment by robertapplin** [comment]
@AnthonyLim23 All the changes have been made, shall I move these changes to the fork, or do you have any more review to do on this?

**Comment by AnthonyLim23** [comment]
I will test it again, so you only need to copy it other once

**Comment by AnthonyLim23** [comment]
the start time is still changing when I turn on simultaneous

File: scripts/test/Muon/CMakeLists.txt
Comment (AnthonyLim23) [inline scripts/test/Muon/CMakeLists.txt:88]: should we delete this?

Comment (robertapplin) [reply scripts/test/Muon/CMakeLists.txt:88]: It is only temporarily disabled, it will be uncommented in a later PR when seq fitting works

**Comment by AnthonyLim23** [comment]
lets merge this into your fork. Could you also rebase the next PR off of this one?

**Comment by robertapplin** [comment]
As agreed, this is now the base branch for this refactor

**Review by StephenSmith25 — DISMISSED** [review]
Looks good, only thing I saw was related to the parameters not staying the same when switching back and forth between workspaces, but I see you've already addressed this in the TF asymmetry PR. Just some conflicts to address.

This is definitely better than what we had before, so really good work there 👍 ! Just one comment: there appears to be a big chain of inheritance going on which can be a little bit challenging to maintain,  as in the end everything is dependent on the `basic_fitting_* `implementation so changes to that file have quite a large knock on effect elsewhere. Not worth holding this up for though, as it is much better than before, just something to keep in mind.

**Comment by robertapplin** [comment]
Thanks for taking a look. I have added a design doc to try explain the composition of these changes in order to avoid the confusion you mentioned as much as possible.

I'll fix the conflicts once we get the other two PRs merged to this one

**Comment by robertapplin** [comment]
Just converting this to a draft to make sure it doesn't get merged yet

**Comment by robertapplin** [comment]
I have fixed the conflicts now so should be ready whenever you are free @StephenSmith25

**Comment by AnthonyLim23** [comment]
I broke it.

Load MUSR 62260-4
Add a Gaus osc, with freq set to 1.3
Set to simultaneous, with freq and sigma shared
Do a sequential fit on all of the data
Set to TF asymmetry
Do another sequential fit
Set back to normal
Add a flat background
Do another sequential fit
Turn TF asymmetry back on
Try t orun another sequential fit - error

**Comment by robertapplin** [comment]
Thanks it should be ready for another test @AnthonyLim23

**Comment by AnthonyLim23** [comment]
the last bug was fixed.

This might be out of scope but is worth considering for the future. 
1. Load HIFI 84447-9
2. Add a fitting function (doesnt matter which)
3. Set to simultaneous
4. Use the `...` to open the extra dialog.
5. Tick the `log` check box (should default to run number)
6. On the parameter value, select `set all to log` and the `fix all`
7. Go to the sequential tab
8. Notice that all of the runs use the same log value instead of their own.

**Comment by AnthonyLim23** [comment]
A cancel option on the sequential tab would be nice

**Comment by AnthonyLim23** [comment]
the other things I have looked at:
- double pulse seems to work
- rebin is used correctly
- updating fitting ranges is carried over

**Comment by robertapplin** [comment]
> the last bug was fixed.
> 
> This might be out of scope but is worth considering for the future.
> 
> 1. Load HIFI 84447-9
> 2. Add a fitting function (doesnt matter which)
> 3. Set to simultaneous
> 4. Use the `...` to open the extra dialog.
> 5. Tick the `log` check box (should default to run number)
> 6. On the parameter value, select `set all to log` and the `fix all`
> 7. Go to the sequential tab
> 8. Notice that all of the runs use the same log value instead of their own.

I agree I think both of these things are out of the scope of this PR as the current version of Mantid is the same

**Review by StephenSmith25 — APPROVED** [review]
Looks good to me. Please wait for Anthony to approve as well before merging.

**Comment by AnthonyLim23** [comment]
I couldnt see anything obvious thats wrong.

**Review by AnthonyLim23 — APPROVED** [review]

**Comment by AnthonyLim23** [comment]
> > the last bug was fixed.
> > This might be out of scope but is worth considering for the future.
> > 
> > 1. Load HIFI 84447-9
> > 2. Add a fitting function (doesnt matter which)
> > 3. Set to simultaneous
> > 4. Use the `...` to open the extra dialog.
> > 5. Tick the `log` check box (should default to run number)
> > 6. On the parameter value, select `set all to log` and the `fix all`
> > 7. Go to the sequential tab
> > 8. Notice that all of the runs use the same log value instead of their own.
> 
> I agree I think both of these things are out of the scope of this PR as the current version of Mantid is the same

seems to be fixed now

**Pull request merged**
